pkg/virtual/apiexport/authorizer/permission_claims_authorizer.go

```diff
@@ -0,0 +1,167 @@
+	listAllAPIBindings func() ([]*apisv1alpha1.APIBinding, error)
+	listAPIBindings    func(clusterName logicalcluster.Name) ([]*apisv1alpha1.APIBinding, error)
+	delegate           authorizer.Authorizer
+}
+
+func (a *permissionClaimsAuthorizer) Authorize(ctx context.Context, attr authorizer.Attributes) (authorized authorizer.Decision, reason string, err error) {
+	if !attr.IsResourceRequest() {
+		return authorization.DelegateAuthorization("non-resource request", a.delegate).Authorize(ctx, attr)
+	}
+
+	clusterName, isWildcardRequest, err := genericapirequest.ClusterNameOrWildcardFrom(ctx)
+	if err != nil {
+		return authorizer.DecisionNoOpinion, "", fmt.Errorf("error getting cluster from request: %w", err)
+	}
+
+	apiDomainKey := dynamiccontext.APIDomainKeyFrom(ctx)
+	parts := strings.Split(string(apiDomainKey), "/")
+	if len(parts) < 2 {
+		return authorizer.DecisionNoOpinion, "", fmt.Errorf("invalid API domain key")
+	}
+
+	claimingAPIExportCluster := parts[0]
+	claimingAPIExportName := parts[1]
+
+	apiExport, err := a.getAPIExport(claimingAPIExportCluster, claimingAPIExportName)
+	if err != nil {
+		return authorizer.DecisionNoOpinion, "", fmt.Errorf("error finding API export %q/%q: %w", claimingAPIExportCluster, claimingAPIExportName, err)
+	}
+
+	// if the requested resource is a bound resource it cannot be claimed.
+	if isBoundResource(attr, apiExport) {
+		return authorization.DelegateAuthorization("bound resource", a.delegate).Authorize(ctx, attr)
+	}
+
+	// Find binding that matches the requested API export in the requested workspace.
+	// If the requested API export has no corresponding API binding then the access must be denied
+	// as the user did not accept any permission claims.
+	allBindings, err := func() ([]*apisv1alpha1.APIBinding, error) {
+		if isWildcardRequest {
+			return a.listAllAPIBindings()
+		}
+		return a.listAPIBindings(clusterName)
+	}()
+	if err != nil {
+		return authorizer.DecisionNoOpinion, "", fmt.Errorf("error finding API bindings for API export %q/%q: %w", claimingAPIExportCluster, claimingAPIExportName, err)
+	}
+	for _, binding := range allBindings {
+		if binding.Spec.Reference.Export == nil {
+			continue
+		}
+
+		if binding.Spec.Reference.Export.Name != apiExport.Name ||
+			binding.Spec.Reference.Export.Path != logicalcluster.From(apiExport).String() {
+			continue
+		}
+
+		for _, claim := range binding.Spec.PermissionClaims {
+			if claim.Group != attr.GetAPIGroup() || claim.Resource != attr.GetResource() {
+				continue
+			}
+			// If at least one claimed resource is not accepted by the user, access must be denied
+			if claim.State != apisv1alpha1.ClaimAccepted {
+				return authorizer.DecisionNoOpinion, fmt.Sprintf("claim %v of API export %q/%q is not accepted", claim, logicalcluster.From(apiExport), apiExport.Name), nil
+			}
+			requestingAllResources := attr.GetName() == "" && attr.GetNamespace() == ""
+			resourceSelectorMatches := claim.All || requestingAllResources
```

Comment:
Do we want to allow this? I thought `claim.All == false && resourceSelector == nil` meant that we still denied it because the user did not make an explicit choice.

Does https://github.com/kcp-dev/kcp/blob/main/pkg/apis/apis/v1alpha1/types_apiexport.go#L199 not prevent this case?

This seems like it has the potential to leak things if a user forgets a field, even if it is better on UX.

Reply:
See this comment when the field was introduced: https://github.com/kcp-dev/kcp/pull/2286#issuecomment-1300653317

Reply:
I guess even I'm not following that rule either 🤦‍♂️  https://github.com/kcp-dev/kcp/pull/2456/files#diff-1ce4a4f7308c17b5ca27f9f57099ae355d2086266eea334dc6f9ae977000c2e6R150